File: android/src/main/kotlin/com/example/easy_native/EasyNativeFlowManager.kt
package com.example.easy_native

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Looper

enum class NativeFlowState {
    IDLE,
    ACTIVE,
    CLOSING,
}

// Coordinates a native flow. It is not a stack manager; Android owns the real Activity stack.
object EasyNativeFlowManager {
    private val trackedActivities: MutableList<Activity> = mutableListOf()
    private val completedRequestIds: MutableSet<String> = mutableSetOf()
    private var state: NativeFlowState = NativeFlowState.IDLE

    fun onActivityCreated(activity: Activity) {
        if (routeNameOf(activity) != null && !trackedActivities.contains(activity)) {
            trackedActivities.add(activity)
            if (state != NativeFlowState.CLOSING) {
                state = NativeFlowState.ACTIVE
            }
        }
    }

    fun onActivityDestroyed(activity: Activity) {
        trackedActivities.remove(activity)
        if (!activity.isChangingConfigurations) {
            val requestId = requestIdOf(activity)
            if (requestId != null) {
                if (completedRequestIds.contains(requestId)) {
                    completedRequestIds.remove(requestId)
                } else {
                    EasyNativePlugin.completeRoute(requestId, null, "nativeDestroy")
                }
            }
        }
        if (trackedActivities.isEmpty()) {
            state = NativeFlowState.IDLE
        }
    }

    fun hasActiveNativeFlow(): Boolean = state != NativeFlowState.IDLE || trackedActivities.isNotEmpty()

    fun push(context: Context, routeName: String, arguments: Any?, requestId: String?): Map<String, Any?> {
        checkCanRoute("nativePush")?.let { return it }
        val intent = EasyNativeRouteRegistry.getNativeIntent(context, routeName, arguments)
            ?: return failure("Native route is not registered: $routeName", "nativePush")
        attachRequestId(intent, requestId)
        val hadActiveFlow = hasActiveNativeFlow()
        val startResult = start(context, intent, "nativePush")
        if (!startResult.success) return startResult.result
        applyPushAnimation(context)
        state = NativeFlowState.ACTIVE
        EasyNativeLogger.log(EasyNativeLogLevel.DEBUG, "native push $routeName")
        return success(if (hadActiveFlow) "nativePush" else "openNativeFlow")
    }

    fun present(context: Context, routeName: String, arguments: Any?, requestId: String?): Map<String, Any?> {
        checkCanRoute("nativePresent")?.let { return it }
        val intent = EasyNativeRouteRegistry.getNativeIntent(context, routeName, arguments)
            ?: return failure("Native route is not registered: $routeName", "nativePresent")
        intent.putExtra(EasyNativeRouteRegistry.EXTRA_PRESENTED, true)
        attachRequestId(intent, requestId)
        val startResult = start(context, intent, "nativePresent")
        if (!startResult.success) return startResult.result
        applyPresentAnimation(context)
        state = NativeFlowState.ACTIVE
        EasyNativeLogger.log(EasyNativeLogLevel.DEBUG, "native present $routeName")
        return success("nativePresent")
    }

    fun replace(
        context: Context,
        routeName: String,
        arguments: Any?,
        result: Any?,
        requestId: String?,
    ): Map<String, Any?> {
        checkCanRoute("nativeReplace")?.let { return it }
        val intent = EasyNativeRouteRegistry.getNativeIntent(context, routeName, arguments)
            ?: return failure("Native route is not registered: $routeName", "nativeReplace")
        attachRequestId(intent, requestId)
        val top = trackedActivities.lastOrNull()
        val startResult = start(context, intent, "nativeReplace")
        if (!startResult.success) return startResult.result
        applyPushAnimation(context)
        state = NativeFlowState.ACTIVE
        top?.let { completeRouteIfNeeded(it, result, "nativeReplace") }
        top?.finish()
        EasyNativeLogger.log(EasyNativeLogLevel.DEBUG, "native replace $routeName")
        return success("nativeReplace")
    }

    fun pop(result: Any?): Map<String, Any?> {
        checkCanRoute("nativePop")?.let { return it }
        val top = trackedActivities.lastOrNull()
            ?: return failure("No active native flow", "nativePop")

        if (trackedActivities.size <= 1) {
            state = NativeFlowState.CLOSING
        }

        completeRouteIfNeeded(top, result, "nativePop")
        top.finish()
        applyCloseAnimation(top)
        EasyNativeLogger.log(EasyNativeLogLevel.DEBUG, "native pop")
        return success("nativePop")
    }

    fun popUntil(routeName: String): Map<String, Any?> {
        checkCanRoute("nativePopUntil")?.let { return it }
        val index = trackedActivities.indexOfLast { routeNameOf(it) == routeName }
        if (index < 0) {
            return failure("Route not found in native stack: $routeName", "nativePopUntil")
        }
        val toFinish = trackedActivities.drop(index + 1)
        toFinish.reversed().forEach {
            completeRouteIfNeeded(it, null, "nativePopUntil")
            it.finish()
            applyCloseAnimation(it)
        }
        EasyNativeLogger.log(EasyNativeLogLevel.DEBUG, "native popUntil $routeName")
        return success("nativePopUntil")
    }

    fun pushAndRemoveUntil(
        context: Context,
        routeName: String,
        arguments: Any?,
        untilRoute: String?,
        requestId: String?,
    ): Map<String, Any?> {
        checkCanRoute("nativePushAndRemoveUntil")?.let { return it }
        val index = untilRoute?.let { target ->
            trackedActivities.indexOfLast { routeNameOf(it) == target }
        } ?: -1
        val toFinish = if (index >= 0) trackedActivities.drop(index + 1) else trackedActivities.toList()
        val intent = EasyNativeRouteRegistry.getNativeIntent(context, routeName, arguments)
            ?: return failure("Native route is not registered: $routeName", "nativePushAndRemoveUntil")
        attachRequestId(intent, requestId)
        val startResult = start(context, intent, "nativePushAndRemoveUntil")
        if (!startResult.success) return startResult.result
        applyPushAnimation(context)
        state = NativeFlowState.ACTIVE
        toFinish.reversed().forEach {
            completeRouteIfNeeded(it, null, "nativePushAndRemoveUntil")
            it.finish()
            applyCloseAnimation(it)
        }
        EasyNativeLogger.log(EasyNativeLogLevel.DEBUG, "native pushAndRemoveUntil $routeName")
        return success("nativePushAndRemoveUntil")
    }

    fun closeAll(result: Any?): Map<String, Any?> {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            return failure("Must be called from main thread", "closeNativeFlow")
        }
        if (trackedActivities.isEmpty()) {
            state = NativeFlowState.IDLE
            return success("noActiveNativeFlow")
        }
        if (state == NativeFlowState.CLOSING) {
            return success("nativeFlowAlreadyClosing")
        }
        state = NativeFlowState.CLOSING
        var deliveredResult = false
        trackedActivities.toList().reversed().forEach { activity ->
            val requestId = requestIdOf(activity)
            completeRouteIfNeeded(
                activity,
                if (!deliveredResult && requestId != null) result else null,
                "closeNativeFlow",
            )
            if (requestId != null) {
                deliveredResult = true
            }
            activity.finish()
            applyCloseAnimation(activity)
        }
        EasyNativeLogger.log(EasyNativeLogLevel.DEBUG, "close native flow")
        return success("closeNativeFlow")
    }

    private fun start(context: Context, intent: Intent, action: String): StartResult {
        return try {
            if (context !is Activity) {
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
            StartResult(true, success(action))
        } catch (throwable: Throwable) {
            EasyNativeLogger.log(EasyNativeLogLevel.ERROR, "startActivity failed", throwable)
            StartResult(false, failure("Failed to start native route: ${throwable.message}", action))
        }
    }

    private data class StartResult(
        val success: Boolean,
        val result: Map<String, Any?>,
    )

    private fun applyPushAnimation(context: Context) {
        (context as? Activity)?.overridePendingTransition(
            R.anim.easy_native_slide_in_right,
            R.anim.easy_native_hold,
        )
    }

    private fun applyPresentAnimation(context: Context) {
        (context as? Activity)?.overridePendingTransition(
            R.anim.easy_native_slide_in_bottom,
            R.anim.easy_native_hold,
        )
    }

    private fun applyCloseAnimation(activity: Activity) {
        if (activity.intent.getBooleanExtra(EasyNativeRouteRegistry.EXTRA_PRESENTED, false)) {
            applyDismissAnimation(activity)
        } else {
            applyPopAnimation(activity)
        }
    }

    private fun applyPopAnimation(activity: Activity) {
        activity.overridePendingTransition(
            R.anim.easy_native_hold,
            R.anim.easy_native_slide_out_right,
        )
    }

    private fun applyDismissAnimation(activity: Activity) {
        activity.overridePendingTransition(
            R.anim.easy_native_hold,
            R.anim.easy_native_slide_out_bottom,
        )
    }

    private fun routeNameOf(activity: Activity): String? {
        return activity.intent.getStringExtra(EasyNativeRouteRegistry.EXTRA_ROUTE_NAME)
    }

    private fun requestIdOf(activity: Activity): String? {
        return activity.intent.getStringExtra(EasyNativeRouteRegistry.EXTRA_REQUEST_ID)
    }

    private fun attachRequestId(intent: Intent, requestId: String?) {
        if (!requestId.isNullOrBlank()) {
            intent.putExtra(EasyNativeRouteRegistry.EXTRA_REQUEST_ID, requestId)
        }
    }

    private fun completeRouteIfNeeded(activity: Activity, result: Any?, action: String) {
        val requestId = requestIdOf(activity) ?: return
        completeRoute(requestId, result, action)
    }

    private fun completeRoute(requestId: String, result: Any?, action: String) {
        if (!completedRequestIds.add(requestId)) return
        EasyNativePlugin.completeRoute(requestId, result, action)
    }

    private fun success(action: String, data: Map<String, Any?> = emptyMap()): Map<String, Any?> {
        return mapOf(
            "success" to true,
            "action" to action,
            "data" to data,
        )
    }

    private fun failure(message: String, action: String): Map<String, Any?> {
        return mapOf(
            "success" to false,
            "action" to action,
            "message" to message,
        )
    }

    private fun checkCanRoute(action: String): Map<String, Any?>? {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            return failure("Must be called from main thread", action)
        }
        if (state == NativeFlowState.CLOSING) {
            EasyNativeLogger.log(EasyNativeLogLevel.WARNING, "reject $action while native flow is closing")
            return failure("Native flow is closing", action)
        }
        return null
    }
}

File: ios/easy_native/Sources/easy_native/EasyNativeFlowManager.swift
import UIKit

private var easyNativeNavigationObserverKey: UInt8 = 0

public enum NativeFlowState {
    case idle
    case active
    case closing
}

public final class EasyNativeNavigationController: UINavigationController {
    public var isPresentedFlow = false

    public override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        if isBeingDismissed || isMovingFromParent {
            EasyNativeFlowManager.shared.clearIfActive(self)
        }
    }
}

private final class EasyNativeNavigationObserver: NSObject, UINavigationControllerDelegate {
    weak var forwardedDelegate: UINavigationControllerDelegate?
    private var knownViewControllers: [UIViewController] = []

    func startObserving(_ navigationController: UINavigationController) {
        if forwardedDelegate == nil,
           let delegate = navigationController.delegate,
           delegate !== self {
            forwardedDelegate = delegate
        }
        knownViewControllers = navigationController.viewControllers
        navigationController.delegate = self
    }

    func updateStack(_ viewControllers: [UIViewController]) {
        knownViewControllers = viewControllers
    }

    func navigationController(
        _ navigationController: UINavigationController,
        willShow viewController: UIViewController,
        animated: Bool
    ) {
        forwardedDelegate?.navigationController?(
            navigationController,
            willShow: viewController,
            animated: animated
        )
    }

    func navigationController(
        _ navigationController: UINavigationController,
        didShow viewController: UIViewController,
        animated: Bool
    ) {
        let current = navigationController.viewControllers
        let popped = knownViewControllers.filter { old in
            !current.contains(where: { $0 === old })
        }

        popped.reversed().forEach {
            EasyNativeFlowManager.shared.completeRouteIfNeeded(
                $0,
                result: nil,
                action: "nativePop"
            )
        }
        EasyNativeFlowManager.shared.handleNavigationStackChanged(
            navigationController
        )
        knownViewControllers = current

        forwardedDelegate?.navigationController?(
            navigationController,
            didShow: viewController,
            animated: animated
        )
    }
}

private extension UINavigationController {
    var easyNativeNavigationObserver: EasyNativeNavigationObserver? {
        get {
            return objc_getAssociatedObject(
                self,
                &easyNativeNavigationObserverKey
            ) as? EasyNativeNavigationObserver
        }
        set {
            objc_setAssociatedObject(
                self,
                &easyNativeNavigationObserverKey,
                newValue,
                .OBJC_ASSOCIATION_RETAIN_NONATOMIC
            )
        }
    }
}

/// Coordinates a native flow. It is not a stack manager; UIKit owns the real stack.
public final class EasyNativeFlowManager {
    public static let shared = EasyNativeFlowManager()

    private weak var flowRootNavigationController: UINavigationController?
    private weak var flowRootViewController: UIViewController?
    private weak var activeNavigationController: UINavigationController?
    private var completedRequestIds: Set<String> = []
    private var state: NativeFlowState = .idle

    private init() {}

    private func validateState() {
        guard state != .idle, let root = flowRootNavigationController else { return }
        if root is EasyNativeNavigationController { return }
        
        if let flowRoot = flowRootViewController {
            if !root.viewControllers.contains(flowRoot) {
                clearState()
            }
        } else {
            clearState()
        }
    }

    private func clearState() {
        flowRootNavigationController = nil
        flowRootViewController = nil
        activeNavigationController = nil
        state = .idle
    }

    public var hasActiveNativeFlow: Bool {
        validateState()
        return state != .idle || flowRootNavigationController != nil
    }

    public func push(
        routeName: String,
        arguments: Any?,
        from rootNav: UINavigationController?,
        requestId: String?
    ) -> [String: Any] {
        validateState()
        if let error = checkCanRoute(action: "nativePush") {
            return error
        }
        guard let viewController = EasyNativeRouteRegistry.shared.makeViewController(
            routeName: routeName,
            arguments: arguments
        ) else {
            return failure("Native route is not registered: \(routeName)", action: "nativePush")
        }
        attachRequestId(requestId, to: viewController)

        if let active = currentNavigationController() {
            EasyNativeLogger.log(.debug, "native push \(routeName)")
            observeNavigationController(active)
            active.pushViewController(viewController, animated: true)
            return success(action: "nativePush")
        }

        guard let rootNav = rootNav else {
            return failure("Root UINavigationController is missing", action: "nativePush")
        }

        flowRootNavigationController = rootNav
        flowRootViewController = viewController
        activeNavigationController = rootNav
        state = .active
        EasyNativeLogger.log(.debug, "open native flow \(routeName)")
        observeNavigationController(rootNav)
        rootNav.pushViewController(viewController, animated: true)
        return success(action: "openNativeFlow")
    }

    public func present(
        routeName: String,
        arguments: Any?,
        from rootNav: UINavigationController?,
        requestId: String?
    ) -> [String: Any] {
        validateState()
        if let error = checkCanRoute(action: "nativePresent") {
            return error
        }
        guard let viewController = EasyNativeRouteRegistry.shared.makeViewController(
            routeName: routeName,
            arguments: arguments
        ) else {
            return failure("Native route is not registered: \(routeName)", action: "nativePresent")
        }
        attachRequestId(requestId, to: viewController)

        let container = EasyNativeNavigationController(rootViewController: viewController)
        container.isNavigationBarHidden = true
        container.isPresentedFlow = true
        container.modalPresentationStyle = .fullScreen
        observeNavigationController(container)

        let presenter = (currentNavigationController() ?? rootNav)?.easyNativeTopMost()
        guard let presenter = presenter else {
            return failure("Presenter is missing", action: "nativePresent")
        }

        if flowRootNavigationController == nil {
            flowRootNavigationController = container
            flowRootViewController = viewController
        }
        activeNavigationController = container
        state = .active
        EasyNativeLogger.log(.debug, "native present \(routeName)")
        presenter.present(container, animated: true)
        return success(action: "nativePresent")
    }

    public func replace(routeName: String, arguments: Any?, result: Any?, requestId: String?) -> [String: Any] {
        validateState()
        if let error = checkCanRoute(action: "nativeReplace") {
            return error
        }
        guard let viewController = EasyNativeRouteRegistry.shared.makeViewController(
            routeName: routeName,
            arguments: arguments
        ) else {
            return failure("Native route is not registered: \(routeName)", action: "nativeReplace")
        }
        attachRequestId(requestId, to: viewController)

        guard let active = currentNavigationController() else {
            return failure("No active native flow", action: "nativeReplace")
        }

        var stack = active.viewControllers
        let oldLast = stack.last
        if stack.isEmpty {
            stack = [viewController]
        } else {
            stack[stack.count - 1] = viewController
        }
        
        if active === flowRootNavigationController, oldLast === flowRootViewController {
            flowRootViewController = viewController
        }
        if let oldLast = oldLast {
            completeRouteIfNeeded(oldLast, result: result, action: "nativeReplace")
        }
        
        observeNavigationController(active)
        active.setViewControllers(stack, animated: true)
        EasyNativeLogger.log(.debug, "native replace \(routeName)")
        return success(action: "nativeReplace")
    }

    public func pop(result: Any?) -> [String: Any] {
        validateState()
        if let error = checkCanRoute(action: "nativePop") {
            return error
        }
        guard let active = currentNavigationController() else {
            return failure("No active native flow", action: "nativePop")
        }

        let isPoppingFlowRoot = (active === flowRootNavigationController && active.viewControllers.last === flowRootViewController)

        if active.viewControllers.count <= 1 || isPoppingFlowRoot {
            if active !== flowRootNavigationController || active is EasyNativeNavigationController {
                active.viewControllers.reversed().enumerated().forEach { index, viewController in
                    completeRouteIfNeeded(
                        viewController,
                        result: index == 0 ? result : nil,
                        action: "nativeDismiss"
                    )
                }
                active.dismiss(animated: true) {
                    self.activeNavigationController = self.currentNavigationController()
                }
                EasyNativeLogger.log(.debug, "native dismiss presented flow")
                return success(action: "nativeDismiss")
            }
            return closeAll(result: result)
        }

        if let top = active.viewControllers.last {
            completeRouteIfNeeded(top, result: result, action: "nativePop")
        }
        observeNavigationController(active)
        active.popViewController(animated: true)
        EasyNativeLogger.log(.debug, "native pop")
        return success(action: "nativePop")
    }

    public func popUntil(routeName: String) -> [String: Any] {
        validateState()
        if let error = checkCanRoute(action: "nativePopUntil") {
            return error
        }
        guard let active = currentNavigationController() else {
            return failure("No active native flow", action: "nativePopUntil")
        }
        guard let target = active.viewControllers.last(where: { $0.easyNativeRouteName == routeName }) else {
            return failure("Route not found in native stack: \(routeName)", action: "nativePopUntil")
        }
        if let index = active.viewControllers.firstIndex(of: target) {
            active.viewControllers.suffix(from: index + 1).forEach {
                completeRouteIfNeeded($0, result: nil, action: "nativePopUntil")
            }
        }
        observeNavigationController(active)
        active.popToViewController(target, animated: true)
        EasyNativeLogger.log(.debug, "native popUntil \(routeName)")
        return success(action: "nativePopUntil")
    }

    public func pushAndRemoveUntil(
        routeName: String,
        arguments: Any?,
        untilRoute: String?,
        requestId: String?
    ) -> [String: Any] {
        validateState()
        if let error = checkCanRoute(action: "nativePushAndRemoveUntil") {
            return error
        }
        guard let viewController = EasyNativeRouteRegistry.shared.makeViewController(
            routeName: routeName,
            arguments: arguments
        ) else {
            return failure("Native route is not registered: \(routeName)", action: "nativePushAndRemoveUntil")
        }
        attachRequestId(requestId, to: viewController)
        guard let active = currentNavigationController() else {
            return failure("No active native flow", action: "nativePushAndRemoveUntil")
        }

        var stack = active.viewControllers
        var removedFlowRoot = false

        if let untilRoute = untilRoute,
           let index = stack.lastIndex(where: { $0.easyNativeRouteName == untilRoute }) {
            let toRemove = stack.suffix(from: index + 1)
            toRemove.forEach {
                completeRouteIfNeeded($0, result: nil, action: "nativePushAndRemoveUntil")
            }
            if toRemove.contains(where: { $0 === flowRootViewController }) {
                removedFlowRoot = true
            }
            stack = Array(stack.prefix(index + 1))
        } else {
            if active === flowRootNavigationController, let flowRoot = flowRootViewController, let index = stack.firstIndex(of: flowRoot) {
                stack.suffix(from: index).forEach {
                    completeRouteIfNeeded($0, result: nil, action: "nativePushAndRemoveUntil")
                }
                stack = Array(stack.prefix(index))
                removedFlowRoot = true
            } else {
                stack.forEach {
                    completeRouteIfNeeded($0, result: nil, action: "nativePushAndRemoveUntil")
                }
                if stack.contains(where: { $0 === flowRootViewController }) {
                    removedFlowRoot = true
                }
                stack.removeAll()
            }
        }
        
        stack.append(viewController)
        
        if active === flowRootNavigationController && removedFlowRoot {
            flowRootViewController = viewController
        }
        
        observeNavigationController(active)
        active.setViewControllers(stack, animated: true)
        EasyNativeLogger.log(.debug, "native pushAndRemoveUntil \(routeName)")
        return success(action: "nativePushAndRemoveUntil")
    }

    public func closeAll(result: Any?) -> [String: Any] {
        validateState()
        guard Thread.isMainThread else {
            return failure("Must be called from main thread", action: "closeNativeFlow")
        }
        guard let root = flowRootNavigationController else {
            state = .idle
            return success(action: "noActiveNativeFlow")
        }
        if state == .closing {
            return success(action: "nativeFlowAlreadyClosing")
        }
        state = .closing
        completeActiveRoutes(result: result, action: "closeNativeFlow")

        let completion = {
            self.flowRootNavigationController = nil
            self.flowRootViewController = nil
            self.activeNavigationController = nil
            self.state = .idle
        }

        if let root = root as? EasyNativeNavigationController,
           root.isPresentedFlow || root.presentingViewController != nil {
            root.dismiss(animated: true, completion: completion)
        } else {
            let popRoot = {
                if let flowRoot = self.flowRootViewController,
                   root.viewControllers.contains(where: { $0 === flowRoot }) {
                    root.popToViewController(flowRoot, animated: false)
                    root.popViewController(animated: true)
                }
                completion()
            }
            if root.presentedViewController != nil {
                root.dismiss(animated: false, completion: popRoot)
            } else {
                popRoot()
            }
        }
        EasyNativeLogger.log(.debug, "close native flow")
        return success(action: "closeNativeFlow")
    }

    public func clearIfActive(_ navigationController: EasyNativeNavigationController) {
        if flowRootNavigationController === navigationController {
            flowRootNavigationController = nil
            flowRootViewController = nil
            activeNavigationController = nil
            state = .idle
        } else if activeNavigationController === navigationController {
            activeNavigationController = currentNavigationController()
        }
    }

    public func handleNavigationStackChanged(_ navigationController: UINavigationController) {
        if navigationController === flowRootNavigationController,
           let flowRoot = flowRootViewController,
           !navigationController.viewControllers.contains(where: { $0 === flowRoot }) {
            clearState()
            return
        }

        if activeNavigationController === navigationController,
           navigationController.viewControllers.isEmpty {
            activeNavigationController = currentNavigationController()
        }
    }

    private func currentNavigationController() -> UINavigationController? {
        guard let root = flowRootNavigationController else {
            return activeNavigationController
        }

        var current: UINavigationController = root
        while let presented = current.presentedViewController as? EasyNativeNavigationController {
            current = presented
        }
        return current
    }

    private func attachRequestId(_ requestId: String?, to viewController: UIViewController) {
        guard let requestId = requestId, !requestId.isEmpty else { return }
        viewController.easyNativeRequestId = requestId
        viewController.easyNativeCompletionToken = EasyNativeRouteCompletionToken(requestId: requestId)
    }

    private func observeNavigationController(_ navigationController: UINavigationController) {
        let observer = navigationController.easyNativeNavigationObserver
            ?? EasyNativeNavigationObserver()
        navigationController.easyNativeNavigationObserver = observer
        observer.startObserving(navigationController)
    }

    public func completeRouteIfNeeded(_ viewController: UIViewController, result: Any?, action: String) {
        guard let requestId = viewController.easyNativeRequestId else { return }
        completeRoute(requestId: requestId, result: result, action: action)
    }

    public func completeRoute(requestId: String, result: Any?, action: String) {
        if action == "nativeDeinit" {
            let wasCompleted = completedRequestIds.contains(requestId)
            completedRequestIds.remove(requestId)
            if wasCompleted {
                return
            }
            EasyNativePlugin.completeRoute(requestId: requestId, result: result, action: action)
            return
        }

        guard !completedRequestIds.contains(requestId) else { return }
        completedRequestIds.insert(requestId)
        EasyNativePlugin.completeRoute(requestId: requestId, result: result, action: action)
    }

    private func completeActiveRoutes(result: Any?, action: String) {
        let controllers = activeFlowViewControllers().reversed()
        var deliveredResult = false
        controllers.forEach { viewController in
            let requestId = viewController.easyNativeRequestId
            completeRouteIfNeeded(
                viewController,
                result: !deliveredResult && requestId != nil ? result : nil,
                action: action
            )
            if requestId != nil {
                deliveredResult = true
            }
        }
    }

    private func activeFlowViewControllers() -> [UIViewController] {
        guard let root = flowRootNavigationController else { return [] }
        var controllers: [UIViewController] = []

        if let flowRoot = flowRootViewController,
           let index = root.viewControllers.firstIndex(of: flowRoot) {
            controllers.append(contentsOf: root.viewControllers.suffix(from: index))
        } else {
            controllers.append(contentsOf: root.viewControllers)
        }

        var current = root.presentedViewController as? EasyNativeNavigationController
        while let nav = current {
            controllers.append(contentsOf: nav.viewControllers)
            current = nav.presentedViewController as? EasyNativeNavigationController
        }

        return controllers
    }

    private func checkCanRoute(action: String) -> [String: Any]? {
        guard Thread.isMainThread else {
            return failure("Must be called from main thread", action: action)
        }
        if state == .closing {
            EasyNativeLogger.log(.warning, "reject \(action) while native flow is closing")
            return failure("Native flow is closing", action: action)
        }
        return nil
    }

    private func success(action: String, data: [String: Any]? = nil) -> [String: Any] {
        return [
            "success": true,
            "action": action,
            "data": data ?? [:],
        ]
    }

    private func failure(_ message: String, action: String) -> [String: Any] {
        return [
            "success": false,
            "action": action,
            "message": message,
        ]
    }
}
